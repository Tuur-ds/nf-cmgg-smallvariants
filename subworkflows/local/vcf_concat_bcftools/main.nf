//
// Concatenate the VCFs back together with bcftools concat
//

include { BCFTOOLS_CONCAT } from '../../../modules/nf-core/bcftools/concat/main'
include { TABIX_TABIX     } from '../../../modules/nf-core/tabix/tabix/main'

workflow VCF_CONCAT_BCFTOOLS {
    take:
        ch_vcfs // channel: [ val(meta), path(vcf), path(tbi) ]
        val_tabix // boolean: whether to create a index or not

    main:

    ch_versions = Channel.empty()

    ch_vcfs
        .map { meta, vcf, tbi ->
            def new_meta = meta + [id:meta.sample ?: meta.family]
            [ groupKey(new_meta, meta.split_count), vcf, tbi ]
        }
        .groupTuple()
        .set { ch_concat_input }

    BCFTOOLS_CONCAT(
        ch_concat_input
    )
    ch_versions = ch_versions.mix(BCFTOOLS_CONCAT.out.versions.first())

    if(val_tabix) {
        TABIX_TABIX(
            BCFTOOLS_CONCAT.out.vcf
        )
        ch_versions = ch_versions.mix(TABIX_TABIX.out.versions.first())

        BCFTOOLS_CONCAT.out.vcf
            .join(TABIX_TABIX.out.tbi, failOnDuplicate: true, failOnMismatch: true)
            .map { meta, vcf, tbi ->
                // Remove the bed counter from the meta field
                def new_meta = meta - meta.subMap("split_count")
                [ new_meta, vcf, tbi ]
            }
            .set { ch_vcf_tbi }
    } else {
        BCFTOOLS_CONCAT.out.vcf
            .map { meta, vcf ->
                // Remove the bed counter from the meta field
                def new_meta = meta - meta.subMap("split_count")
                [ new_meta, vcf ]
            }
            .set { ch_vcf_tbi }
    }

    emit:
    vcfs = ch_vcf_tbi       // channel: [ val(meta), path(vcf), path(tbi) ]

    versions = ch_versions  // channel: [ versions.yml ]
}