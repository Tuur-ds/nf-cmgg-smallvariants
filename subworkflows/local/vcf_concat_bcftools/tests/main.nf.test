nextflow_workflow {

    name "Test Workflow VCF_CONCAT_BCFTOOLS"
    script "../main.nf"
    workflow "VCF_CONCAT_BCFTOOLS"

    tag "subworkflows"
    tag "subworkflows_local"
    tag "vcf_concat_bcftools"

    test("vcf_concat_bcftools - no_tabix") {

        config "./nextflow.config"

        when {
            workflow {
                """
                input[0] = Channel.of([
                    [id:"NA24143", sample:"NA24143", family:"Ashkenazim", family_samples:"NA24143", caller:"haplotypecaller", split_count:2],
                    file(params.gvcf1, checkIfExists:true),
                    file(params.gtbi1, checkIfExists:true)
                ],[
                    [id:"NA24143", sample:"NA24143", family:"Ashkenazim", family_samples:"NA24143", caller:"haplotypecaller", split_count:2],
                    file(params.vcf1, checkIfExists:true),
                    file(params.tbi1, checkIfExists:true)
                ])
                input[1] = false
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(
                    workflow.out.vcfs.collect { [it[0], "variantsMD5:${path(it[1]).vcf.variantsMD5}"] }
                ).match() }
            )
        }

    }

    test("vcf_concat_bcftools - tabix") {

        config "./nextflow.config"

        when {
            workflow {
                """
                input[0] = Channel.of([
                    [id:"NA24143", sample:"NA24143", family:"Ashkenazim", family_samples:"NA24143", caller:"haplotypecaller", split_count:2],
                    file(params.gvcf1, checkIfExists:true),
                    file(params.gtbi1, checkIfExists:true)
                ],[
                    [id:"NA24143", sample:"NA24143", family:"Ashkenazim", family_samples:"NA24143", caller:"haplotypecaller", split_count:2],
                    file(params.vcf1, checkIfExists:true),
                    file(params.tbi1, checkIfExists:true)
                ])
                input[1] = true
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(
                    workflow.out.vcfs.collect { [it[0], "variantsMD5:${path(it[1]).vcf.variantsMD5}", it[2][-10..-1]] }
                ).match() }
            )
        }

    }

}
