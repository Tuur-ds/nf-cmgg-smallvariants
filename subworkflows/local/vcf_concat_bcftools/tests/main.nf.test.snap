{
    "vcf_concat_bcftools - tabix": {
        "content": [
            [
                [
                    {
                        "id": "NA24143",
                        "sample": "NA24143",
                        "family": "Ashkenazim",
                        "family_samples": "NA24143",
                        "caller": "haplotypecaller"
                    },
                    "variantsMD5:843352db8fe3f441ffa026dc72a30c35",
                    "vcf.gz.tbi"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.9.1",
            "nextflow": "24.10.0"
        },
        "timestamp": "2024-11-14T13:41:32.794902067"
    },
    "vcf_concat_bcftools - no_tabix": {
        "content": [
            [
                [
                    {
                        "id": "NA24143",
                        "sample": "NA24143",
                        "family": "Ashkenazim",
                        "family_samples": "NA24143",
                        "caller": "haplotypecaller"
                    },
                    "variantsMD5:843352db8fe3f441ffa026dc72a30c35"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.9.1",
            "nextflow": "24.10.0"
        },
        "timestamp": "2024-11-14T13:42:04.623824246"
    }
}